lms/views/decorators/h_api.py

```diff
@@ -122,3 +124,124 @@ def wrapper(request, jwt):  # pylint: disable=too-many-branches
+    """
+    Create a group in h if one doesn't exist already.
+
+    Call the h API to create a group for the LTI course, if one doesn't exist
+    already.
+
+    Use this function as a decorator rather than calling it directly.
+    The wrapped view must take ``request`` and ``jwt`` arguments::
+
+      @view_config(...)
+      @create_h_group
+      def my_view(request, jwt):
+          ...
+
+    """
+    # FIXME: This function doesn't do anything with the ``jwt`` argument,
+    # other than pass it through to the wrapped view (or to the next decorator
+    # on the wrapped view). The ``jwt`` argument has to be here because:
+    #
+    # - This decorator is called by the ``@lti_launch`` decorator (because
+    #   ``@lti_launch`` is always placed above this decorator on decorated views)
+    #   and ``@lti_launch`` passes a ``jwt_token`` argument when it calls this
+    #   decorator.
+    #
+    # - The views that this decorator decorates expect a ``jwt`` argument so this
+    #   decorator has to pass it to them (or rather it has to pass it down to the
+    #   next decorator down in the stack, and it eventually gets passed to the
+    #   view).
+    #
+    # This should all be refactored so that views and view decorators aren't
+    # tightly coupled and arguments don't need to be passed through multiple
+    # decorators to the view.
+    def wrapper(request, jwt):  # pylint: disable=too-many-branches
+        # Only create users for application instances that we've enabled the
+        # auto provisioning features for.
+        try:
+            oauth_consumer_key = request.params["oauth_consumer_key"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "oauth_consumer_key" missing from LTI params')
+        enabled_consumer_keys = request.registry.settings["auto_provisioning"]
+        if oauth_consumer_key not in enabled_consumer_keys:
+            return wrapped(request, jwt)
+
+        try:
+            tool_consumer_instance_guid = request.params["tool_consumer_instance_guid"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "tool_consumer_instance_guid" missing from LTI params')
+
+        try:
+            context_id = request.params["context_id"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "context_id" missing from LTI params')
+
+        # Show the user an error page if the group hasn't been created yet and
+        # the user isn't allowed to create the group because they aren't an
+        # administrator, instructor or teaching assistant.
+        try:
+            roles = request.params["roles"].lower()
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "roles" missing from LTI params')
+        if not any(role in roles for role in ("administrator", "instructor", "teachingassisstant")):
+            raise HTTPBadRequest('Teacher must configure assignment first')
```

Comment:
This check for whether the user is an administrator, instructor or teaching assistant (as opposed to a "learner" or something else) is also duplicated twice now: <https://github.com/hypothesis/lms/blob/947a51874f21ba43d65899d1cd652e8bc4e90371/lms/views/lti_launches.py#L24-L32>.

Seems like another thing that belongs in a service, but perhaps for now I can find somewhere (e.g. in `util`) to move it.

Reply:
Pedantic request: change "teacher" to "instructor".

Reply:
Ugh, I think I'm just gonna leave this duplicated for now, rather than make up another random util function for it. Fix it properly with a service or request property later